File: sim/chip.go
package sim

import (
	"fmt"
	"math"
	"sort"
	"sync"

	"github.com/but80/fmfm.core/ymf/ymfdata"
)

// Chip は、FM音源チップ全体を表す型です。
type Chip struct {
	Mutex sync.Mutex
	// sampleRate は、出力波形の目標サンプルレートです。
	sampleRate float64
	// totalLevel は、出力のトータルな音量[dB]です。
	totalLevel float64
	// dumpMIDIChannel は、ダンプ表示対象のMIDIチャンネルです。未使用時は -1 です。
	dumpMIDIChannel int
	// channels は、このチップが備える全チャンネルです。
	channels []*Channel

	currentOutput []float64
}

// NewChip は、新しい Chip を作成します。
func NewChip(sampleRate, totalLevel float64, dumpMIDIChannel int) *Chip {
	chip := &Chip{
		sampleRate:      sampleRate,
		totalLevel:      totalLevel,
		dumpMIDIChannel: dumpMIDIChannel,
		channels:        make([]*Channel, ymfdata.ChannelCount),
		currentOutput:   make([]float64, 2),
	}
	chip.initChannels()
	return chip
}

var debugDumpCount = 0

// Next は、次のサンプルを生成し、その左右それぞれの振幅を返します。
func (chip *Chip) Next() (float64, float64) {
	var l, r float64
	for _, channel := range chip.channels {
		chip.Mutex.Lock()
		cl, cr := channel.next()
		chip.Mutex.Unlock()
		l += cl
		r += cr
	}
	v := math.Pow(10, chip.totalLevel/20)

	if 0 <= chip.dumpMIDIChannel {
		debugDumpCount++
		if int(chip.sampleRate/ymfdata.DebugDumpFPS) <= debugDumpCount {
			debugDumpCount = 0
			toDump := []*Channel{}
			for _, ch := range chip.channels {
				if ch.midiChannelID == chip.dumpMIDIChannel && epsilon < ch.currentLevel() {
					toDump = append(toDump, ch)
				}
			}
			if 0 < len(toDump) {
				sort.Slice(toDump, func(i, j int) bool {
					return toDump[i].currentLevel() < toDump[j].currentLevel()
				})
				for _, ch := range toDump {
					fmt.Print(ch.dump())
				}
				fmt.Println("------------------------------")
			}
		}
	}

	return l * v, r * v
}

func (chip *Chip) initChannels() {
	chip.channels = make([]*Channel, ymfdata.ChannelCount)
	for i := range chip.channels {
		chip.channels[i] = newChannel(i, chip)
	}
}

File: cpp/sim/chip.cpp
#pragma once
#include "./chip.h"

namespace sim {



// NewChip は、新しい Chip を作成します。
Chip *NewChip(float64 sampleRate, float64 totalLevel, int dumpMIDIChannel) {
	auto chip = __ptr((const Chip){
		sampleRate: sampleRate,
		totalLevel: totalLevel,
		dumpMIDIChannel: dumpMIDIChannel,
		channels: make([41mexpr[0m[31m<*ast.ArrayType>(&ast.ArrayType{Lbrack:11070, Len:ast.Expr(nil), Elt:(*ast.StarExpr)(0xc4200f9780)})[0m, ymfdata->ChannelCount),
		currentOutput: make([41mexpr[0m[31m<*ast.ArrayType>(&ast.ArrayType{Lbrack:11129, Len:ast.Expr(nil), Elt:(*ast.Ident)(0xc4200f9860)})[0m, 2),
	});
	ChipPtr__initChannels(chip);
	return chip;
}

auto debugDumpCount = 0;

// Next は、次のサンプルを生成し、その左右それぞれの振幅を返します。
MULTIRESULT ChipPtr__Next(Chip *chip) {
	float64 l, float64 r;
	for (int _ = 0; _ < (int)chip->channels.size(); _++) {
		auto channel = chip->channels[_];
		sync::Mutex__Lock(chip->Mutex);
		auto cl, cr = ChannelPtr__next(channel);
		sync::Mutex__Unlock(chip->Mutex);
		l = cl;
		r = cr;
	}
	auto v = math::Pow(10, chip->totalLevel/20);
	if (0 <= chip->dumpMIDIChannel) {
		debugDumpCount++
		if (int(chip->sampleRate/ymfdata->DebugDumpFPS) <= debugDumpCount) {
			debugDumpCount = 0;
			auto toDump = {};
			for (int _ = 0; _ < (int)chip->channels.size(); _++) {
				auto ch = chip->channels[_];
				if (ch->midiChannelID == chip->dumpMIDIChannel && epsilon < ChannelPtr__currentLevel(ch)) {
					toDump = append(toDump, ch);
				}
			}
			if (0 < len(toDump)) {
				sort::Slice(toDump, 				bool UNKNOWN(int i, int j) {
					return [41mobjectOf[0m[31m<*ast.IndexExpr>(&ast.IndexExpr{X:(*ast.Ident)(0xc4200fe740), Lbrack:11960, Index:(*ast.Ident)(0xc4200fe760), Rbrack:11962})[0m[41mobjectOf[0m[31m<*ast.IndexExpr>(&ast.IndexExpr{X:(*ast.Ident)(0xc4200fe740), Lbrack:11960, Index:(*ast.Ident)(0xc4200fe760), Rbrack:11962})[0mChannelPtr__currentLevel(toDump[i]) < [41mobjectOf[0m[31m<*ast.IndexExpr>(&ast.IndexExpr{X:(*ast.Ident)(0xc4200fe7c0), Lbrack:11987, Index:(*ast.Ident)(0xc4200fe7e0), Rbrack:11989})[0m[41mobjectOf[0m[31m<*ast.IndexExpr>(&ast.IndexExpr{X:(*ast.Ident)(0xc4200fe7c0), Lbrack:11987, Index:(*ast.Ident)(0xc4200fe7e0), Rbrack:11989})[0mChannelPtr__currentLevel(toDump[j]);
				});
				for (int _ = 0; _ < (int)toDump.size(); _++) {
					auto ch = toDump[_];
					fmt::Print(ChannelPtr__dump(ch));
				}
				fmt::Println("------------------------------");
			}
		}
	}
	return l*v, r*v;
}

void ChipPtr__initChannels(Chip *chip) {
	chip->channels = make([41mexpr[0m[31m<*ast.ArrayType>(&ast.ArrayType{Lbrack:12221, Len:ast.Expr(nil), Elt:(*ast.StarExpr)(0xc4200feca0)})[0m, ymfdata->ChannelCount);
	for (int i = 0; i < (int)chip->channels.size(); i++) {
		chip->channels[i] = newChannel(i, chip);
	}
}

}
